### PR TITLE
feat: add funnel 

### PR DESCRIPTION
**Pull request by huxiaoyun**

### CODE REVIEW

File: src/component/Funnel.js [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+    if (hide || !trapezoids || !trapezoids.length) {
+      return null;
+    }
+
+    const layerClass = classNames('recharts-trapezoids', className);
+
+    return (
+      <Layer className={layerClass}>
+        {this.renderTrapezoids()}
+        {label && this.renderLabels(trapezoids)}
```
Comment (xile611) [inline src/component/Funnel.js]: use `<LabelList />` to render labels

File: src/component/Funnel.js [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+                const interpolatorY = interpolateNumber(prev.y, entry.y);
+
+                return {
+                  ...entry,
+                  x: interpolatorX(t),
+                  y: interpolatorY(t),
+                };
+              }
+
+              const interpolatorX = interpolateNumber(0, entry.x);
```
Comment (xile611) [inline src/component/Funnel.js]: It's strange to only interpolate `x` and `y`

File: src/component/Funnel.js [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+      const tooltipPosition = {
+        x: x + upperWidth / 2,
+        y: y + rowHeight / 2,
+      };
+
+      return {
+        x,
+        y,
+        upperWidth,
+        lowerWidth,
```
Comment (xile611) [inline src/component/Funnel.js]: add a props width to describe the view box of each trapezoid

```js
width: Math.max(upperWidth, lowerWidth),
```

File: src/component/Funnel.js (hunk 1 above)
Comment (xile611) [reply src/component/Funnel.js]:
```jsx
{(!isAnimationActive || isAnimationFinished) &&
          LabelList.renderCallByParent(this.props, trapezoids)}
```

File: demo/component/Trapezoid.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import { Surface, Trapezoid } from 'recharts';
+
+export default class Demo extends Component {
+
+  static displayName = 'TrapezoidDemo';
+
+  render() {
+    return (
+      <Surface width={500} height={1000}>
+        <Trapezoid
```
Comment (xile611) [inline demo/component/Trapezoid.js]: Add demo of animated <Trapezoid />

File: demo/component/FunnelChart.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  static displayName = 'FunnelChartDemo';
+
+  state = initialState;
+
+  render() {
+    return (
+      <div className="funnel-charts">
+        <p>Simple FunnelChart</p>
+        <div className="funnel-chart-wrapper" style={{ width: '50%', height: '300', backgroundColor: '#f5f5f5' }}>
+          <ResponsiveContainer>
```
Comment (xile611) [inline demo/component/FunnelChart.js]: Add demo of FunnelChart when data changes

File: src/shape/Trapezoid.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+            d={getTrapezoidPath(x, y, upperWidth, lowerWidth, height)}
+          />
+        </g>
+      );
+    }
+
+    return (
+      <Animate
+        canBegin={totalLength > 0}
+        from={{ upperWidth, height, x, y }}
```
Comment (xile611) [inline src/shape/Trapezoid.js]: lowerWidth needs to be added to the animation props.

File: src/component/Funnel.js [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+    ]),
+    id: PropTypes.string,
+  };
+
+  static defaultProps = {
+    stroke: '#fff',
+    fill: '#808080',
+    legendType: 'rect',
+    // The abscissa of pole
+    cx: '0%',
```
Comment (xile611) [inline src/component/Funnel.js]: This props is unused

File: src/component/Funnel.js [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+    // The ordinate of pole
+    cy: '0%',
```
Comment (xile611) [inline src/component/Funnel.js]: This props is unused

File: src/component/Funnel.js [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+    labelLine: true,
```
Comment (xile611) [inline src/component/Funnel.js]: This props is unused

File: src/component/Funnel.js [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+    animationDuration: PropTypes.number,
+    animationEasing: PropTypes.oneOf([
+      'ease',
+      'ease-in',
+      'ease-out',
+      'ease-in-out',
+      'spring',
+      'linear',
+    ]),
+    id: PropTypes.string,
```
Comment (xile611) [inline src/component/Funnel.js]: This props is unused

File: src/chart/Treemap.js [hunk 11]
```diff
@@ -207,6 +214,8 @@ class Treemap extends Component {
       PropTypes.arrayOf(PropTypes.node),
       PropTypes.node,
     ]),
+    type: PropTypes.string,
+    nestIndexContent: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
```
[the remark this answers is not in the document]
Comment (huxiaoyun) [reply src/chart/Treemap.js]: image
use this api to customize nest index content。

File: src/chart/Treemap.js [hunk 12]
```diff
@@ -346,30 +483,61 @@ class Treemap extends Component {
...
   }
 
-  static renderContentItem(content, nodeProps) {
+  static renderContentItem(content, nodeProps, type) {
     if (React.isValidElement(content)) {
       return React.cloneElement(content, nodeProps);
     } if (_.isFunction(content)) {
       return content(nodeProps);
     }
+    // 加强默认形状
```
[the remark this answers is not in the document]
Comment (huxiaoyun) [reply src/chart/Treemap.js]: done

**Pull request merged**